File: openHAB/SegmentedUITableViewCell.m
//
//  SegmentedUITableViewCell.m
//  openHAB
//
//  Created by Victor Belov on 17/01/14.
//  Copyright (c) 2014 Victor Belov. All rights reserved.
//

#import "SegmentedUITableViewCell.h"
#import "OpenHABWidget.h"
#import "OpenHABItem.h"
#import "OpenHABWidgetMapping.h"

@implementation SegmentedUITableViewCell
@synthesize widgetSegmentedControl;

- (id)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        self.widgetSegmentedControl = (UISegmentedControl *)[self viewWithTag:500];
        self.selectionStyle = UITableViewCellSelectionStyleNone;
        self.separatorInset = UIEdgeInsetsZero;
    }
    return self;
}

- (void)displayWidget
{
    self.textLabel.text = [self.widget labelText];
    widgetSegmentedControl.apportionsSegmentWidthsByContent = YES;
    [self.widgetSegmentedControl removeAllSegments];
    [self.widgetSegmentedControl setApportionsSegmentWidthsByContent:YES];
    for (OpenHABWidgetMapping *mapping in self.widget.mappings) {
        [self.widgetSegmentedControl insertSegmentWithTitle:mapping.label atIndex:[self.widget.mappings indexOfObject:mapping] animated:NO];
    }
    [self.widgetSegmentedControl setSelectedSegmentIndex:[self.widget mappingIndexByCommand:self.widget.item.state]];
    [self.widgetSegmentedControl addTarget:self
                                    action:@selector(pickOne:)
                          forControlEvents:UIControlEventValueChanged];
}

-(void) pickOne:(id)sender
{
    UISegmentedControl *segmentedControl = (UISegmentedControl *)sender;
    NSLog(@"Segment pressed %d", [segmentedControl selectedSegmentIndex]);
    if (self.widget.mappings != nil) {
        OpenHABWidgetMapping *mapping = [self.widget.mappings objectAtIndex:[segmentedControl selectedSegmentIndex]];
        [self.widget sendCommand:mapping.command];
    }
}

@end

File: openHAB/SegmentedUITableViewCell.swift
// Copyright (c) 2010-2023 Contributors to the openHAB project
//
// See the NOTICE file(s) distributed with this work for additional
// information.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0
//
// SPDX-License-Identifier: EPL-2.0

import os.log
import UIKit

class SegmentedUITableViewCell: GenericUITableViewCell {
    private let feedbackGenerator = UIImpactFeedbackGenerator(style: .light)

    // @IBOutlet private var customTextLabel: UILabel!
    @IBOutlet private var widgetSegmentControl: UISegmentedControl!

    required init?(coder: NSCoder) {
        super.init(coder: coder)

        selectionStyle = .none
        separatorInset = .zero
    }

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)

        selectionStyle = .none
        separatorInset = .zero
    }

    override func displayWidget() {
        customTextLabel?.text = widget.labelText
        customDetailTextLabel?.text = widget.labelValue ?? ""

        widgetSegmentControl.apportionsSegmentWidthsByContent = true
        widgetSegmentControl.removeAllSegments()
        widgetSegmentControl.apportionsSegmentWidthsByContent = true

        for (index, mapping) in (widget?.mappingsOrItemOptions ?? []).enumerated() {
            widgetSegmentControl.insertSegment(withTitle: mapping.label, at: index, animated: false)
        }

        widgetSegmentControl.isMomentary = widget.mappingsOrItemOptions.count == 1 || widget.item?.state == "NULL"
        widgetSegmentControl.selectedSegmentIndex = widgetSegmentControl.isMomentary ? -1 : Int(widget.mappingIndex(byCommand: widget.item?.state) ?? -1)
        widgetSegmentControl.addTarget(self, action: #selector(SegmentedUITableViewCell.pickOne(_:)), for: .valueChanged)
    }

    @objc
    func pickOne(_ sender: Any?) {
        guard let segmentedControl = sender as? UISegmentedControl, let mapping = widget.mappingsOrItemOptions[safe: segmentedControl.selectedSegmentIndex] else {
            return
        }

        os_log("Segment pressed %d", log: .default, type: .info, segmentedControl.selectedSegmentIndex)
        widget.sendCommand(mapping.command)
        feedbackGenerator.impactOccurred()
    }
}
